ios/engine/KMEI/KeymanEngineDemo/MainViewController.swift

```diff
@@ -0,0 +1,315 @@
+    setTextFieldStyle(textView2)
+    setTextViewStyle(textView3)
+
+    textView1.setKeymanDelegate(self)
+    textView2.setKeymanDelegate(self)
+    textView2.placeholder = "text field"
+    textView2.borderStyle = .roundedRect
+    textView2.clearButtonMode = .whileEditing
+
+    textView1.frame = CGRect(x: margin, y: keymanLabel.frame.maxY + 5.0,
+                             width: contentView.frame.width - margin * 2.0, height: 80.0)
+    textView2.frame = CGRect(x: margin, y: textView1.frame.maxY + margin,
+                             width: contentView.frame.width - margin * 2.0, height: 30.0)
+    normalLabel.frame = CGRect(x: margin, y: textView2.frame.maxY + margin,
+                               width: contentWidth - margin * 2.0, height: 40.0)
+    textView3.frame = CGRect(x: margin, y: normalLabel.frame.maxY + margin,
+                             width: contentWidth - margin * 2.0, height: 40.0)
+
+    textView1.viewController = self
+    textView2.viewController = self
+    contentView.addSubview(textView1)
+    contentView.addSubview(textView2)
+    contentView.addSubview(textView3)
+
+    // Button - keyboard picker
+    let kbButton = KeyboardPickerButton(presentingVC: self)
+    kbButton.center = CGPoint(x: contentWidth / 2.0, y: textView3.frame.maxY + 30.0)
+    kbButton.frame = kbButton.frame.integral
+    kbButton.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin]
+    contentView.addSubview(kbButton)
+
+    // Button - close keyboard
+    let dismissButton = UIButton(type: .roundedRect)
+    dismissButton.setTitle("Close Keyboard", for: .normal)
+    dismissButton.addTarget(self, action: #selector(self.dismissKeyboard), for: .touchUpInside)
+    dismissButton.sizeToFit()
+    dismissButton.center = CGPoint(x: contentWidth / 2.0, y: kbButton.frame.maxY + 30.0)
+    dismissButton.frame = dismissButton.frame.integral
+    dismissButton.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin]
+    contentView.addSubview(dismissButton)
+
+    // Button - download custom keyboard
+    let downloadButton = UIButton(type: .roundedRect)
+    downloadButton.setTitle("Download custom keyboard", for: .normal)
+    downloadButton.addTarget(self, action: #selector(self.downloadButtonTapped), for: .touchUpInside)
+    downloadButton.sizeToFit()
+    downloadButton.center = CGPoint(x: contentWidth / 2.0, y: dismissButton.frame.maxY + 30.0)
+    downloadButton.frame = downloadButton.frame.integral
+    downloadButton.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin]
+    contentView.addSubview(downloadButton)
+
+    // Content view setup
+    contentView.frame = CGRect(x: 0.0, y: 0.0, width: contentWidth, height: downloadButton.frame.maxY + margin)
+    contentView.contentSize = contentView.frame.size
+    contentView.backgroundColor = UIColor.gray
+    contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+    return contentView
+  }
+
+  @objc func dismissKeyboard() {
```

Comment:
There seem to be a number of methods that went missing from this file... but I did notice that it's also part of the new 'Demo' project section, since some division there was found to be necessary during the Swift conversion, if I recall.

I assume these methods exist elsewhere now within the Engine itself, rather than in this class?  Any chance you could point me to them?

Reply:
I'm not sure which methods you're referring to here. But the engine no longer has a `MainViewController`. That was moved to the Demo target along with all other code specific to the Demo app and unnecessary in the actual engine library. If you look at `Keyman.h`, `MainViewController.h` is not imported.